docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,414 @@
+10 shares. One year later, 1 share is worth 120 tokens, so if I transfer
+1,000 tokens, the program converts and asks to move 8.33 shares. The same concept
+applies to mints and burns.
+
+Following this model, we never have to worry about compounding interest. We
+simply need to call the right program to perform instructions.
+
+## Required Changes
+
+A new interest-bearing token program is only one component of the overall solution,
+which also comprises an interface for tokens (i-tokens) to be easily integrated in
+on-chain programs, web apps, and wallets.
+
+### Token Program Conformance
+
+Since our solution entails the creation of a new token program, we're opening the
+floodgates for other app developers to do the same. To ensure safe and easy integration
+of new token programs, we need a set of conformance tests based on the current
+SPL token test suite.
+
+#### Instructions
+
+A new token program must include all of the same instructions as SPL token,
+but the "unchecked" variants should return errors. Concretely, this means the
+program must implement:
+
+* `InitializeMint`
+* `InitializeHolding` (previously `InitializeAccount`)
+* `Revoke`
+* `SetAuthority`
+* `CloseHolding` (previously `CloseAccount`)
+* `FreezeHolding` (previously `FreezeAccount`)
+* `ThawHolding` (previously `ThawAccount`)
+* `TransferChecked`
+* `ApproveChecked`
+* `MintToChecked`
+* `BurnChecked`
+* `InitializeHoldingWithOwnerAddress` (previously `InitializeAccount2`)
+
+And the program should throw errors for:
+
+* `Transfer`
+* `Approve`
+* `MintTo`
+* `Burn`
+* `InitializeMultisig` (use a more general multisig)
+
+New instructions required:
+
+* `CreateHolding`: only performs `Allocate` and `Assign` to self,
+useful for creating a holding when you don't know the program that you're
+interacting with. See the Associated Token Account section for how to use this.
+
+There are also new read-only instructions, which write data to a provided account
+buffer:
+
+* `AmountToUiAmount`: convert the given share amount to a UI token amount
+* `UiAmountToAmount`: convert the given UI token amount to an exact share amount
+
+See the Runtime section for more information on how these are used.
+
+Programs and wallets that wish to support i-tokens must update the instructions
+to use all `Checked` variants and use the new read-only instructions to fetch
+information.
+
+#### Conversions and Wrapper Library
+
+The SPL token library provides wrappers compatible with any conforming program.
+The wrapper simply checks the program id and deserializes the account data
+as an `spl_token::Holding` or `spl_token::Mint`. If a client program requires access
+to data specific to an i-token program, it needs to use that program's serializers.
+
+To convert between amount and UI amount, the wrapper library calls the `AmountToUiAmount`
+instruction on the i-token program, or for SPL token accounts, it calls
+`spl_token::amount_to_ui_amount` inline.
```

Comment:
just fyi, I leaned that wstETH (wrapper version of stETH) just exists for the distinction of value-accruing vs re-basing token differences to be used for separately depending on upper defi primitive's token accounting assumption: https://help.lido.fi/en/articles/5231836-what-is-wrapped-steth-wsteth

Also, I think we now have return data runtime mechanism. :)